subgraph/src/balances.ts

```diff
@@ -1,79 +1,222 @@
 
-  // APR
-  market.lastTotalSupply = market.lastTotalSupply.minus(previousBalance).plus(newUnderlyingSupplyBalance);
-  market.supplyIndex = newSupplyIndex;
-  market.supplyUpdateBlockTimestamp = blockTimestamp;
-  market.save();
+  balance.save();
 
   return balance;
 };
 
+/**
+ * blockTimestamp & borrowUpdateBlockTimestamp must be in
+ * the same version to distribute rewards correctly
+ */
+const computeBorrowIndexUpdate = (
+  marketAddress: Address,
+  blockTimestamp: BigInt,
+  indexUnits: u8,
+  balance: Balance,
+  market: Market,
+  skipV1Accounting: boolean
+): void => {
+  const newBorrowIndex = updateBorrowIndex(marketAddress, blockTimestamp);
+
+  const newBorrowIndexOnPool = updateBorrowIndexOnPool(marketAddress, blockTimestamp, indexUnits);
+  const newBorrowIndexInP2P = updateBorrowIndexInP2P(marketAddress, blockTimestamp, indexUnits);
+
+  const accumulatedMorphoV1 = accrueMorphoTokens(
+    newBorrowIndex,
+    balance.userBorrowIndex,
+    balance.underlyingBorrowBalance
+  );
+  if (!skipV1Accounting) {
+    // We skip accounting for a Borrower position update on V1 script
+
+    balance.userBorrowIndex = newBorrowIndex;
+    balance.accumulatedBorrowMorphoV1 = balance.accumulatedBorrowMorphoV1.plus(accumulatedMorphoV1);
+    market.borrowIndex = newBorrowIndex;
+    market.borrowUpdateBlockTimestampV1 = blockTimestamp;
+  }
+
+  const accumulatedMorphoV2 = accrueMorphoTokens(
+    newBorrowIndexOnPool,
+    balance.userBorrowOnPoolIndex,
+    balance.scaledBorrowOnPool
+  ).plus(accrueMorphoTokens(newBorrowIndexInP2P, balance.userBorrowInP2PIndex, balance.scaledBorrowInP2P));
+
+  balance.accumulatedBorrowMorphoV2 = balance.accumulatedBorrowMorphoV2.plus(accumulatedMorphoV2);
+
+  if (blockTimestamp.le(VERSION_2_TIMESTAMP)) {
+    if (!skipV1Accounting) balance.accumulatedBorrowMorpho = balance.accumulatedBorrowMorpho.plus(accumulatedMorphoV1);
+  } else {
+    balance.accumulatedBorrowMorpho = balance.accumulatedBorrowMorpho.plus(accumulatedMorphoV2);
+  }
+  balance.userBorrowOnPoolIndex = newBorrowIndexOnPool;
+  balance.userBorrowInP2PIndex = newBorrowIndexInP2P;
+
+  market.poolBorrowIndex = newBorrowIndexOnPool;
+  market.p2pBorrowIndex = newBorrowIndexInP2P;
+
+  market.borrowUpdateBlockTimestamp = blockTimestamp;
+  market.save();
+  balance.save();
+};
 export const updateBorrowBalanceAndMarket = (
   marketAddress: Address,
   userAddress: Address,
   blockNumber: BigInt,
   blockTimestamp: BigInt,
   newBalanceP2P: BigInt,
-  newBalanceOnPool: BigInt
+  newBalanceOnPool: BigInt,
+  indexUnits: u8,
+  skipV1Accounting: boolean
 ): Balance => {
   const market = getOrInitMarket(marketAddress, blockTimestamp);
   const balance = getOrInitBalance(userAddress, marketAddress, blockTimestamp);
+  if (market.borrowUpdateBlockTimestamp.lt(VERSION_2_TIMESTAMP) && blockTimestamp.ge(VERSION_2_TIMESTAMP)) {
+    computeBorrowIndexUpdate(marketAddress, VERSION_2_TIMESTAMP, indexUnits, balance, market, false);
+  }
```

Comment:
Nice!

Reply:
To be consistent, I think we should compare `blockTimestamp.gt(VERSION_2_TIMESTAMP)` otherwise we could compute it twice if `blockTimestamp == VERSION_2_TIMESTAMP`

Implications being obviously ridiculous, the only motivation would be consistency

Reply:
but the second one is not going to be computed since delta timestamp is zero, but I agree w you that is going to be cleaner this way (less computation)